### ISSUE TITLE
Out-of-box LuaSocket support

### ISSUE THREAD

**Issue by nevkontakte** [X-Friese/FlyWithLua#11]
Being able to make network connections and HTTP requests would be very helpful for many applications. I'm personally interested in developing some tools for FSEconomy, but I'm sure there are many more use cases.

From what I can tell, [LuaSocket](https://github.com/diegonehab/luasocket) is the most common library for networking in Lua and it is a requirement for many other libraries.

I found some reports of varying recency [[1](https://forums.x-plane.org/index.php?/forums/topic/138672-making-http-requests/&tab=comments#comment-1328904), [2](https://forums.x-plane.org/index.php?/forums/topic/127045-solved-downloading-web-pages-in-fwl/&tab=comments#comment-1221409)] which laim a success. There are other reports that LuaSocket won't work with FlyWithLua [[3](https://forums.x-plane.org/index.php?/forums/topic/71174-luasocket-for-flywithlua/&tab=comments#comment-949529), [4](https://forums.x-plane.org/index.php?/forums/topic/89042-lua-socket/&do=findComment&comment=951602)] for one reason or another. In particular, according to [[5](https://forums.x-plane.org/index.php?/forums/topic/89042-lua-socket/&tab=comments#comment-949526)] FlyWithLua uses a [patched version of LuaJit](http://www.xsquawkbox.net/xpsdk/mediawiki/LuaJIT), which is incompatible with luasocket (not sure why exactly).

Since many of these posts are rather old, I have a few questions:

 1. Are there any reason why including LuaSocket into FlyWithLua (perhaps in Complete Edition only) would be a bad idea? If not, I'd be interested in working on this.
 2. Does FlyWithLua still use that patched LuaJit version? In this repo I was able to find a few already compiled binaries, so it's hard for me to tell.
 3. It seems FlyWithLua NG supports only 64 bit versions of X-Plane, is that right? If so, I'd need to worry only about one flavor of binaries, which would be nice.

Hope some FlyWithLua experts would be able to provide some answers (-:

**Comment by sparker256** [comment]
1) I did look at this quite a while ago but did not get very far. I do think this would be a good idea if possible.

2) FlyWithLua does use a patched version of LuaJit because of this. https://developer.x-plane.com/article/luajit/

3) You are correct that FlyWithLua NG is only 64bit and any updates will be the same.

It is important that we maintain support for all three platforms.

Creating a fork to see what might be possible would be the correct approach.

**Comment by nevkontakte** [comment]
Thanks for the answers (that was rather quick!), I'll try some experiments and report back.

Starting off with a fork seems reasonable, but I wanted to make sure that the whole idea makes sense and get any extra pointers I've missed, which might help me along the way.

https://developer.x-plane.com/article/luajit/ seems to be a copy of http://www.xsquawkbox.net/xpsdk/mediawiki/LuaJIT. Unfortunately, links to both precompiled binaries and sources are broken on both pages. Is https://github.com/X-Friese/FlyWithLua/blob/master/LuaJIT/LuaJIT-2.0.1.zip the patched version of LuaJit used by FlyWithLua?

**Comment by nevkontakte** [comment]
Ok, this does seem to be the patched version of LuaJit, and the patch itself is this [x-plane-luajit.diff.txt](https://github.com/X-Friese/FlyWithLua/files/2364495/x-plane-luajit.diff.txt). If my reading is correct, the patch just disables one function in the public API: `LUA_API lua_State *lua_newstate(lua_Alloc f, void *ud)`.

The good news is, this patch doesn't break binary compatibility, so any lua extension which works with a vanilla LuaJit will work with the customized version used by FlyWithLua.

Sidenote: I stumbled upon another similar patch, which is arguably more future-proof, in FlyWithLua will be upgrading to a new LuaJit version: https://gist.github.com/nddrylliog/8722197.

**Comment by nevkontakte** [comment]
Ok, here are some good news and questions:

I've built the latest luasocket from https://github.com/diegonehab/luasocket against vanilla Lua 5.1 and it worked. This is pretty awesome, because any pre-built lua 5.1 module should just work with FlyWithLua assuming correct file locations.

For posterity, putting contents of [luasocket_add.zip](https://github.com/X-Friese/FlyWithLua/files/2364596/luasocket_add.zip) into "Modules" directory the script below into "Scripts" will make happy little "Hello world!" appear in dev console.

```lua
local http = require('socket.http')
print(http.request("http://urlecho.appspot.com/echo?body=Hello%20world!"))
```
I've tested it on Windows and I'm pretty sure it'll work on linux too, I'll try it a bit later. Unfortunately, I don't have a Mac, so I can only guess.

Now, the questions:

 1. Is it possible to put both Mac and Linux libraries in the Modules directory? They both have .so extension, so I worry there will be a name collision. I doubt it is possible to produce an .so functional on both Mac and Linux at the same time.
 2. In terms of adding those libraries to FlyWithLua repo, what is the right location? Do we want to just stick pre-compiled binaries in or should I try to integrate it with FlyWithLua build scripts?

**Comment by sparker256** [comment]
Here is what I would suggest.

`FlyWithLua/Modules/ltn12.lua
FlyWithLua/Modules/socket.lua
FlyWithLua/Modules/socket/core_64.dll
FlyWithLua/Modules/socket/core_64.so_mac
FlyWithLua/Modules/socket/core_64.so_lin
FlyWithLua/Modules/socket/http.lua
FlyWithLua/Modules/socket/tp.lua
FlyWithLua/Modules/socket/ftp.lua
FlyWithLua/Modules/socket/smtp.lua
FlyWithLua/Modules/socket/url.lua
FlyWithLua/Modules/mime.lua
FlyWithLua/Modules/mime/core_64.dll
FlyWithLua/Modules/mime/core_64.so_mac
FlyWithLua/Modules/mime/core_64.so_lin`

**Comment by nevkontakte** [comment]
Ok, I'll do that. Although I'm not sure if lua loader would actually pick up those "so_lin" and "so_mac" extensions. I'll have to try and see, I guess.

In the meantime I've done some research on whether I can avoid doing cross-compilation myself and found two solutions: http://ulua.io/ and https://luapower.com/, both provide pre-compiled modules for a large set of lua modules. So far seems to be a better option, as they seem to be more up-to-date.

Below goes a huge scope-creep, feel free to ignore: it would be pretty cool to integrate FlyWithLua with some package manager, such as aforementioned ulua.io. But that goes against the idea of using Lua for simple tiny helper scripts, rather than full-blown development environment.

**Comment by sparker256** [comment]
No it will not pick up the so_lin or so_mac.

My thought was that the user would remove the _lin or _mac depending on what platform they are on.

**Comment by sparker256** [comment]
On Linux using your test script and using the pre-compiled from  https://luapower.com/ I am getting this error.

Element no. 1 is: error loading module 'socket.core' from file 'Resources/plugins/FlyWithLua/Modules/socket/core_64.so':
	Resources/plugins/FlyWithLua/Modules/socket/core_64.so: undefined symbol: lua_gettop

Any thoughts?

**Comment by sparker256** [comment]
More testing and on Windows it does work but does not work on my Mac and my Linux box. Since I build all three platforms on my Linux box I am now wondering what might be different that makes this happen.

**Comment by nevkontakte** [comment]
Interesting... Is this the only symbol it complains about?

It'll be a couple of days before I get my linux machine to run X-Plane. In the meantime, could you please try doing the same with ulua.io binaries? You can get them here http://pkg.ulua.io/pkg/socket/3.0rc1-203 (add .zip extension after downloading).

Searching in Google suggests that it most likely means a mismatch between lua version used in the app and one use to compile the extension. So we might have to cross-compile after all...

**Comment by nevkontakte** [comment]
Also some thoughts on potential extension conflict. I think we could customize [this script](https://github.com/X-Friese/FlyWithLua/blob/master/FlyWithLua/Internals/FlyWithLua.ini#L17) a bit to look for different library extensions for Mac and Linux, making it zero-effort for users. 

Looking closer at modules directories, I see these files: `libLuaXML_lib_64.so  libLuaXML_lib.so  LuaXML_lib_64.so  LuaXML_lib.so`. Given that `ldd` on my linux server can only recognize libLuaXML_lib_64.so, I assume that Linux would look for files prefixed with "lib" and Mac would look for a file without prefix. If that's the case, there is no need to customize extensions. Perhaps @X-Friese could confirm this :-)

**Comment by sparker256** [comment]
I did try the ulua and still getting the error on Linux.

I also built using  "-Wl,-export-dynamic" and still getting the same error.

Since I was static linking luajit I tried dynamic linking but still get the error.

**Comment by nevkontakte** [comment]
TL;DR: I give up on idea of using pre-built lua extensions and will try to link them statically in.

Ok, I finally got my linux laptop to run x-plane (boy, those amazing 4 fps...) and had some time to dig into the issue.

The reason why it doesn't work unlike on Windows seems to be the way POSIX systems handle symbol resolution for dynamically loaded libraries. Normally the app itself would embed (or link against) lua and provide those symbols, which is what all Lua infrastructure is feared towards. Lua modules don't link against liblua.so to ensure they only use the implementation provided by host and only it.

In our case FlyWithLua itself is a dynamically loaded plugin, so rules on how symbols provided by it are exported are different. Even if the exports it (which I managed to achieve), when X-Plane loads it with dlopen(..., RTLD_LOCAL) no symbols are made available to subsequently loaded libraries. On Windows it probably uses different visibility logic, which is why it worked (which I find more surprising now).

I see multiple avenues of solving this. To achieve full compatibility with arbitrary lua modules we could employ various hacks, for example https://luarocks.org/modules/siffiejoe/sofix. Which it would be cool in theory, it adds to a risk of creating conflicts with other Lua plugins and is not worth it.

A more practical solution would be to statically link all binary libraries we wish to support and make them available by adding to package.preload table. This does require rebuilding FlyWithLua for every additional module, but this reduces chances of weird conflicts with other plugins and crashes.

Lastly, I'm a bit puzzles how luaxml works. FlyWithLua links against it at build time (as a shared library) but then it does try to load it at dynamically too.
